### ISSUE TITLE
Add RDS support for dbengine Aurora

### ISSUE THREAD

**Issue by apotek** [ansible/ansible#22617]
_From @davidecerri on March 3, 2016 17:40_

##### Issue Type:
- Feature Idea

##### COMPONENT NAME
rds
[module: cloud/amazon/rds.py]

##### Ansible Version:

N/A
##### Summary:

Add support of aws aurora. 
[https://aws.amazon.com/rds/aurora/](https://aws.amazon.com/rds/aurora/)

Aurora is already supported in the latest boto version:
[http://boto3.readthedocs.org/en/latest/reference/services/rds.html?highlight=aurora](http://boto3.readthedocs.org/en/latest/reference/services/rds.html?highlight=aurora)
but the rds module for ansible does not contain aurora as an option:
db_engine:
    description:
      - The type of database.  Used only when command=create.
    required: false
    default: null
    choices: [ 'MySQL', 'oracle-se1', 'oracle-se', 'oracle-ee', 'sqlserver-ee', 'sqlserver-se', 'sqlserver-ex', 'sqlserver-web', 'postgres']

_Copied from original issue: ansible/ansible-modules-core#3180_

**Comment by apotek** [comment]
_From @aioue on April 5, 2016 23:32_

Possibly already fixed in https://github.com/ansible/ansible-modules-core/pull/2646

**Comment by apotek** [comment]
_From @berinle on April 13, 2016 18:40_

@aioue No that is not a fix for this feature request. #2646 is to allow for creating parameter groups for aurora engine type. The main rds module @ https://github.com/h1ppo/ansible-modules-core/blob/cc0ee4ad40acd418f39459e5c5051a5f6ea7686d/cloud/amazon/rds.py#L45 is still not allowing Aurora as an engine type.

On that note, there also needs to be a way to create and customer database cluster group parameters, which I believe should fall under a different ticket, though it's related to this feature request

**Comment by apotek** [comment]
_From @gretel on April 29, 2016 21:50_

that'd be great for `2.1`

**Comment by apotek** [comment]
_From @einarc on June 4, 2016 2:41_

+1, not in doc yet so I assume is not in 2.1, which I'm using.

**Comment by apotek** [comment]
_From @einarc on June 4, 2016 2:42_

@aioue Why is it not in the documentation?

**Comment by apotek** [comment]
_From @JoeyG1973 on June 8, 2016 15:26_

+1 on getting this implemented.

**Comment by apotek** [comment]
_From @kaikousa on July 22, 2016 9:16_

For some reason aurora is listed as a possible value for `db_engine`-option in the documentation but it doesn't allow using it. This happens with `ansible 2.1.0.0`.

**Comment by apotek** [comment]
_From @ansibot on July 31, 2016 17:39_

@bpennypacker, @willthames, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by apotek** [comment]
_From @kaikousa on August 8, 2016 12:39_

I tried copy-pasting the module from devel-branch to my Ansible project's library-folder and running it but got the following error: 

`fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to create instance: Standalone DB Instances are not supported for the specified engine"}`

It seems that even with the Aurora support added to devel, the rds module is not fit for supporting Aurora. boto3 has different API for creating the Aurora-cluster. I will take a run at hacking up a custom module and see where it takes me :)

**Comment by apotek** [comment]
_From @naslanidis on August 14, 2016 3:47_

@kaikousa 

I get a similar issue with this module:

When I run it with multi_zone set to yes:
 "msg": "Failed to create instance: VPC Multi-AZ DB Instances are not available for engine: aurora"}

When I run it with multi_zone set to no: 
"msg": "Failed to create instance: Standalone DB Instances are not supported for the specified engine"}

So something is clearly not right here. I very much need support for Aurora but a new module might have to be created. This one needs to be migrated to Boto3 as 2 is not longer supported and lacks a huge amount of new functionality. Personally I think this module handles too much and could be broken down into smaller sections using boto3.

**Comment by apotek** [comment]
_From @naeemhaq on August 25, 2016 17:36_

+10

**Comment by apotek** [comment]
_From @ansibot on August 25, 2016 17:45_

@bpennypacker, @willthames, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by apotek** [comment]
_From @naslanidis on August 25, 2016 22:27_

Just an FYI, if you're seeing this:

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Failed to create instance: Standalone DB Instances are not supported for the specified engine"}

It's because you need to have an Aurora cluster created before you try and create the Aurora instance and you must pass the cluster name as one of the arguments when calling the create_db_instance API action. 

I have written a new version of the RDS module that uses the updated API (boto 3) that I'm using at the moment. I will look to share that in the next few days but as there is an existing module so I didn't want to tread on any toes. 

I've also written an rds_cluster module in boto3 that manages the Aurora clusters. I will share that in the coming days as well as a new PR.

**Comment by apotek** [comment]
_From @naslanidis on August 29, 2016 1:3_

FYI I've added PR for a new module to manage RDS clusters for Aurora if anyone wants to use/test. 

https://github.com/ansible/ansible-modules-extras/pull/2804

**Comment by apotek** [comment]
_From @ienliven on September 1, 2016 0:58_

rds_cluster works great, well done! how do I create the actual instance inside the cluster ? rds module does not seem to support it. I am getting "Failed to create instance: Standalone DB Instances are not supported for the specified engine"

**Comment by apotek** [comment]
_From @naslanidis on September 1, 2016 1:10_

@ienliven 

At least when using the boto3 api, there's an argument called 'DBClusterIdentifier' where you pass the name of your cluster when creating the instance. That's what it's complaining about when it says standalone isn't supported, it needs to know which cluster to join.

https://boto3.readthedocs.io/en/latest/reference/services/rds.html#RDS.Client.create_db_instance

The old rds_module for creating instances uses the older API and doesn't implement this.  I have rewritten that module in boto3 and have been using it for a few weeks. I wasn't going to submit it as a new PR as I was going to get in touch with the original author if possible and see if they'd like to work with me to support boto3. However I will upload a copy of it to my forked repo and you can add it to your library and use it/test it if you like.

**Comment by apotek** [comment]
_From @naslanidis on September 1, 2016 1:41_

@ienliven 

https://github.com/naslanidis/ansible-modules-extras/blob/updated_rds_boto3_module/cloud/amazon/rds_boto3.py

**Comment by apotek** [comment]
_From @ienliven on September 1, 2016 3:56_

@naslanidis You rock! Thanks

**Comment by apotek** [comment]
_From @ansibot on September 12, 2016 16:58_

@bpennypacker, @willthames, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by apotek** [comment]
_From @ansibot on October 6, 2016 14:57_

@bpennypacker, @willthames, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by apotek** [comment]
_From @willthames on October 21, 2016 7:40_

@naslanidis I am one of the (inactive) current maintainers of the rds module, and I'd be happy to get a PR in that implements boto3 while maintaining boto2.rds2 compatiblity, if possible.

**Comment by apotek** [comment]
_From @willthames on October 21, 2016 7:40_

needs_contributor

**Comment by apotek** [comment]
_From @willthames on October 21, 2016 7:45_

Relies on #5343

**Comment by apotek** [comment]
_From @naslanidis on October 21, 2016 8:37_

Hi @willthames 

I needed an updated version of this module a little while ago so I've finished it for boto3. 

https://github.com/ansible/ansible-modules-extras/pull/2845

I don't mind collaborating to merge it all into one module that supports boto 2 and 3. I purposely reused as much as possible and kept it in the same style in case of that exact possibility.  In the Google groups it was mentioned that boto3 rewrites of existing modules would be accepted from 2.3 so I didn't worry about continuing to support boto2.

**Comment by apotek** [comment]
_From @willthames on October 21, 2016 9:31_

@naslanidis if it's feature compatible after a rewrite to boto3, and we can get agreement on #2845 (or just point to the list discussion), then I'd be happy for the boto3 version to replace the boto2 version in 2.3. Given your contribution it would likely make sense for you to be a maintainer for the module if you can spare the time (I just posted a request for new maintainers on ansible-devel)

**Comment by apotek** [comment]
_From @naslanidis on October 30, 2016 2:6_

Hi @willthames 

I'd be willing to be a maintainer for the boto3 version of this module and can spare the time. I've not done much with earlier boto versions so it would be more difficult for me to maintain backwards compatibility but if that's not such a concern in Ansible 2.3 onwards then I'd be well suited to assisting.

**Comment by apotek** [comment]
_From @ansibot on December 7, 2016 18:28_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by kwidholm-tm** [comment]
As instructed at bottom of https://github.com/ansible/ansible-modules-core/issues/3180, I have moved this to ansible/ansible

**Comment by kjartansible** [comment]
+1

**Comment by kclausen-tm** [comment]
+1

**Comment by apotek** [comment]
As instructed at bottom of ansible/ansible-modules-core#3180, I have moved this to ansible/ansible

**Comment by willthames** [comment]
I'm not really working on fixing issues with existing rds. 

If you're able to test the new rds_instance module, that would be greatly appreciated

https://github.com/ansible/ansible/pull/19772

I can't recall at this point whether it has cluster support, but I'll endeavour to add it if not.

**Comment by ansibot** [comment]
@apotek Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by kevinkamps** [comment]
+1

Any idea when this feature will be released?

**Comment by ahmedsajid** [comment]
+1

**Comment by Akasurde** [comment]
@willthames @s-hertel I see Aurora in Ansible RDS docs. If this is true, can we close this issue ?

**Comment by logbue-dataxu** [comment]
This still seems to be an issue. 
Please can someone let us know the status of Ansible support for 
a) AWS RDS Aurora (MySQL; dbengine aurora) and 
b) AWS RDS Aurora PostgreSQL (dbengine aurora-postgresql)
Thanks

**Comment by s-hertel** [comment]
I think this should be left open. Some clarification on what the rds module does and doesn't do:

Can use aurora with `command:` `facts`/`reboot`/`delete`
Cannot (but should be able) with `command:` `create`/`modify`/`snapshot`/`restore`
And then `replicate` and `promote` don't work (as intended) because there are no stand-alone aurora instances.
